packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts

```
@@ -0,0 +1,660 @@
+ * https://github.com/angular/angular-cli/blob/b3f84cc5184337666ce61c07b7b9df418030106f/packages/ngtools/webpack/src/transformers/ctor-parameters.ts#L323-L325
+ * The trick is not ideal though as it preserves the full import (as outlined before), and it
+ * results in a slow-down due to the type checker being involved multiple times. The CLI
+ * worked around this import preserving issue by having another complex post-process step that
+ * detects and elides unused imports. Note that these unused imports could cause unused chunks
+ * being generated by Webpack if the application or library is not marked as side-effect free.
+ *
+ * This is not ideal though, as we basically re-implement the complex import usage resolution
+ * from TypeScript. We can do better by letting TypeScript do the import eliding, but providing
+ * information about the alias declarations (e.g. import specifiers) that should not be elided
+ * because they are actually referenced (as they will now appear in static properties).
+ *
+ * More information about these limitations with transformers can be found in:
+ *   1. https://github.com/Microsoft/TypeScript/issues/17552.
+ *   2. https://github.com/microsoft/TypeScript/issues/17516.
+ *   3. https://github.com/angular/tsickle/issues/635.
+ *
+ * The patch we apply to tell TypeScript about actual referenced aliases (i.e. imported symbols),
+ * matches conceptually with the logic that runs internally in TypeScript when the
+ * `emitDecoratorMetadata` flag is enabled. TypeScript basically surfaces the same problem and
+ * solves it conceptually the same way, but obviously doesn't need to access an `@internal` API.
+ *
+ * See:
+ * https://sourcegraph.com/github.com/microsoft/TypeScript@3eaa7c65f6f076a08a5f7f1946fd0df7c7430259/-/blob/src/compiler/checker.ts#L31219-31257
+ *
+ * @returns A boolean indicating whether the alias reference solution could be patched properly.
+ */
+function patchAliasReferenceResolution(
```

1) Move this function into a separate file, so it's not lost in the rest of the transform code.
2) Instead of returning `boolean`, let's have it throw the error itself. We can rename it to `patchAliasReferenceResolutionOrDie`.
3) The error should be user-friendly and should explain that the version of TS is incompatible with the Angular compiler and they should downgrade (presumably, they're seeing it for the first time after updating TS).